src/formats/jpeg.zig

```diff
@@ -216,11 +223,59 @@ pub const JPEG = struct {
         return result;
     }
 
-    fn writeImage(allocator: std.mem.Allocator, write_stream: *io.WriteStream, image: Image, encoder_options: Image.EncoderOptions) Image.WriteError!void {
-        _ = allocator;
-        _ = write_stream;
-        _ = image;
-        _ = encoder_options;
+    fn writeImage(
+        allocator: std.mem.Allocator,
+        write_stream: *io.WriteStream,
+        image: Image,
+        encoder_options: Image.EncoderOptions,
+    ) Image.WriteError!void {
+        // Extract JPEG-specific encoder options
+        const jpeg_options = encoder_options.jpeg;
+
+        // Validate quality parameter
+        if (jpeg_options.quality == 0 or jpeg_options.quality > 100) {
+            return Image.WriteError.InvalidData;
+        }
+
+        // Determine if format conversion is needed
+        var converted_image = image;
+        var needs_cleanup = false;
+        defer if (needs_cleanup) converted_image.pixels.deinit(allocator);
+
+        const current_format = std.meta.activeTag(image.pixels);
+        const target_format: PixelFormat = switch (current_format) {
+            .grayscale8, .rgb24 => current_format,
+            // Grayscale variants convert to grayscale8
+            .grayscale1, .grayscale2, .grayscale4, .grayscale16, .grayscale8Alpha, .grayscale16Alpha => .grayscale8,
+            // Everything else converts to rgb24
+            else => .rgb24,
+        };
+
+        // Convert format if needed
+        if (target_format != current_format) {
```

Comment:
The conversion should only be done if the user requests it in the encoder options.